File: src/leetcode/preorder_traversal.py
#! /usr/bin/python
#
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

    # Build binary search tree with a list of values
    def into(self, vals):
        for v in vals:
            self.insert(v)

    def insert(self, val):
        if val <= self.val:
            if self.left: self.left.insert(val)
            else        : self.left = TreeNode(val)
        elif val > self.val:
            if self.right: self.right.insert(val)
            else         : self.right = TreeNode(val)

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def preorderTraversal(self, root):
        if root is None: return []
        result = []
        stack  = []
        stack.append(root)
        while stack:
            x = stack.pop()
            result.append(x.val)
            if x.right: stack.append(x.right)
            if  x.left: stack.append(x.left)
        return result

    # recursively traverse the tree
    def recurPreTraverse(self, list, node):
        if node:
            list.append(node.val)
            self.recurPreTraverse(list, node.left)
            self.recurPreTraverse(list, node.right)

File: src/leetcode/PreorderTraversal.java
package LeetCode;

import java.util.List;
import java.util.ArrayList;
import java.util.Stack;

public class PreorderTraversal {

    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<Integer>();
        if (root == null) return result;
        Stack<TreeNode> st = new Stack<TreeNode>();
        st.push(root);
        while (!st.empty()) {
            TreeNode x = st.pop();
            result.add(x.val);
            if (x.right != null) st.push(x.right);
            if (x.left  != null) st.push(x.left);
        }
        return result;
    }
}
